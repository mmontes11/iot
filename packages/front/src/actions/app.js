import {
  TOGGLE_HAMBURGER_MENU,
  TOGGLE_MAP_DIALOG,
  TOGGLE_LANGUAGE_SELECTOR,
  SET_SHOW_ERROR,
} from "constants/actionTypes/app";

export const toggleHamburgerMenu = () => dispatch => {
  dispatch({ type: TOGGLE_HAMBURGER_MENU });
};

export const toggleMapDialog = () => dispatch => {
  dispatch({ type: TOGGLE_MAP_DIALOG });
};

export const toggleLanguageSelector = () => dispatch => {
  dispatch({ type: TOGGLE_LANGUAGE_SELECTOR });
};

export const setShowError = showError => dispatch => {
  dispatch({ type: SET_SHOW_ERROR, showError });
};
