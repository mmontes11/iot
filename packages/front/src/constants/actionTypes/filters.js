export const FILTER_TYPE_SELECT = "FILTER_TYPE_SELECT";
export const ADD_FILTER_TYPE = "ADD_FILTER_TYPE";
export const DELETE_FILTER_TYPE = "DELETE_FILTER_TYPE";
