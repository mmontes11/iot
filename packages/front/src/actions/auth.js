import {
  IS_AUTH,
  SET_USERNAME,
  SET_PASSWORD,
  LOGIN_REQUEST,
  LOGIN_REQUEST_SUCCESS,
  LOGIN_REQUEST_ERROR,
} from "constants/actionTypes/auth";
import { setShowError } from "actions/app";
import iotClient from "lib/iotClient";

export const isAuth = () => dispatch => {
  iotClient.authService
    .isAuth()
    .then(isAuthVal => dispatch({ type: IS_AUTH, isAuth: isAuthVal }))
    .catch(() => dispatch({ type: IS_AUTH, isAuth: false }));
};

export const setUsername = username => dispatch => {
  dispatch({ type: SET_USERNAME, username });
};

export const setPassword = password => dispatch => {
  dispatch({ type: SET_PASSWORD, password });
};

export const login = () => (dispatch, getState) => {
  dispatch({ type: LOGIN_REQUEST });
  const { username, password } = getState().auth;
  iotClient.authService.setCredentials(username, password);
  iotClient.authService
    .getToken()
    .then(response => {
      dispatch({
        type: LOGIN_REQUEST_SUCCESS,
        statusCode: response.statusCode,
        error: null,
      });
      isAuth()(dispatch);
    })
    .catch(error => {
      dispatch({
        type: LOGIN_REQUEST_ERROR,
        statusCode: error.statusCode,
        error,
      });
      setShowError(true)(dispatch);
    });
};

export const logout = () => dispatch => {
  iotClient.authService
    .logout()
    .then(() => {
      isAuth()(dispatch);
    })
    .catch(() => {
      isAuth()(dispatch);
    });
};
