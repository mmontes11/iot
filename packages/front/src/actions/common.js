import { RESET } from "constants/actionTypes/common";

export const reset = preserveError => dispatch => {
  dispatch({ type: RESET, preserveError });
};
