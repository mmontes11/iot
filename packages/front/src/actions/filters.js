import { FILTER_TYPE_SELECT, ADD_FILTER_TYPE, DELETE_FILTER_TYPE } from "constants/actionTypes/filters";

export const selectFilterType = () => dispatch => {
  dispatch({ type: FILTER_TYPE_SELECT });
};

export const addFilterType = addedFilterType => dispatch => {
  dispatch({ type: ADD_FILTER_TYPE, addedFilterType });
};

export const deleteFilterType = deletedFilterType => dispatch => {
  dispatch({ type: DELETE_FILTER_TYPE, deletedFilterType });
};
