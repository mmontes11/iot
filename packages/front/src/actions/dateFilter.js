import {
  DATE_FILTER_SELECTOR_TOGGLE,
  DATE_FILTER_TIME_PERIOD_SELECT,
  DATE_FILTER_TIME_PERIOD_UPDATED,
  DATE_FILTER_TIME_PERIODS_REQUEST,
  DATE_FILTER_TIME_PERIODS_REQUEST_SUCCESS,
  DATE_FILTER_TIME_PERIODS_REQUEST_ERROR,
  DATE_FILTER_TIME_PERIODS_UPDATED,
  DATE_FILTER_START_DATE_UPDATED,
  DATE_FILTER_END_DATE_UPDATED,
} from "constants/actionTypes/dateFilter";
import { ADD_FILTER_TYPE } from "constants/actionTypes/filters";
import { DATE_FILTER_TYPE } from "constants/filterTypes";
import iotClient from "lib/iotClient";

export const toggleDateFilterType = () => dispatch => {
  dispatch({ type: DATE_FILTER_SELECTOR_TOGGLE });
};

export const selectTimePeriod = isActive => dispatch => {
  dispatch({ type: DATE_FILTER_TIME_PERIOD_SELECT });
  if (!isActive) {
    dispatch({ type: DATE_FILTER_TIME_PERIODS_REQUEST });
    iotClient.timePeriodsService
      .getSupportedTimePeriods()
      .then(({ body: { timePeriods } }) => {
        dispatch({ type: DATE_FILTER_TIME_PERIODS_REQUEST_SUCCESS });
        dispatch({ type: DATE_FILTER_TIME_PERIODS_UPDATED, timePeriods });
      })
      .catch(error => {
        dispatch({ type: DATE_FILTER_TIME_PERIODS_REQUEST_ERROR, statusCode: error.statusCode, error });
      });
  }
};

export const updateTimePeriod = updatedTimePeriod => dispatch => {
  dispatch({ type: DATE_FILTER_TIME_PERIOD_UPDATED, updatedTimePeriod });
};

export const updateStartDate = updatedStartDate => dispatch => {
  dispatch({ type: DATE_FILTER_START_DATE_UPDATED, updatedStartDate });
};

export const updateEndDate = updatedEndDate => dispatch => {
  dispatch({ type: DATE_FILTER_END_DATE_UPDATED, updatedEndDate });
};

export const addTimePeriodFilter = updatedTimePeriod => dispatch => {
  dispatch({ type: ADD_FILTER_TYPE, addedFilterType: DATE_FILTER_TYPE });
  dispatch({ type: DATE_FILTER_TIME_PERIOD_UPDATED, updatedTimePeriod });
};

export const addCustomTimePeriodFilter = (updatedStartDate, updatedEndDate) => dispatch => {
  dispatch({ type: ADD_FILTER_TYPE, addedFilterType: DATE_FILTER_TYPE });
  const startDate = new Date(updatedStartDate);
  const endDate = new Date(updatedEndDate);
  if (!Number.isNaN(startDate) || !Number.isNaN(endDate)) {
    dispatch({ type: DATE_FILTER_SELECTOR_TOGGLE, isCustomSelected: true });
  }
  if (!Number.isNaN(startDate)) {
    dispatch({ type: DATE_FILTER_START_DATE_UPDATED, updatedStartDate: startDate });
  }
  if (!Number.isNaN(endDate)) {
    dispatch({ type: DATE_FILTER_END_DATE_UPDATED, updatedEndDate: endDate });
  }
};
