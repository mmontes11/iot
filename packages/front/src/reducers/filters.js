import { FILTER_TYPE_SELECT, ADD_FILTER_TYPE, DELETE_FILTER_TYPE } from "constants/actionTypes/filters";
import { RESET } from "constants/actionTypes/common";
import { THING_FILTER_TYPE, DATE_FILTER_TYPE, FILTER_TYPES } from "constants/filterTypes";
import thingFilterReducer, { initialState as thingFilterInitialState } from "reducers/thingFilter";
import dateFilterReducer, { initialState as dateFilterInitialState } from "reducers/dateFilter";

export const initialState = {
  type: {
    items: FILTER_TYPES,
    isActive: false,
    isDisabled: false,
  },
  thingFilter: thingFilterInitialState,
  dateFilter: dateFilterInitialState,
  items: [],
};

export default (state = initialState, action) => {
  switch (action.type) {
    case FILTER_TYPE_SELECT:
      return { ...state, type: { ...state.type, isActive: !state.type.isActive } };
    case ADD_FILTER_TYPE: {
      const { addedFilterType } = action;
      const addedItems = [...state.items];
      if (!addedItems.includes(addedFilterType)) {
        addedItems.push(addedFilterType);
      }
      const filterTypeItems = state.type.items.filter(item => !addedItems.includes(item));
      const allItemsSelected = FILTER_TYPES.every(item => addedItems.includes(item));
      return {
        ...state,
        type: {
          ...state.type,
          items: filterTypeItems,
          isActive: false,
          isDisabled: allItemsSelected,
        },
        items: addedItems,
      };
    }
    case DELETE_FILTER_TYPE: {
      const { deletedFilterType } = action;
      const addedItems = [...state.items];
      const newItems = addedItems.filter(item => item !== deletedFilterType);
      const filterTypeItems = [...state.type.items];
      if (!filterTypeItems.includes(deletedFilterType)) {
        filterTypeItems.push(deletedFilterType);
      }
      const allItemsSelected = FILTER_TYPES.every(item => newItems.includes(item));
      const nextState = {
        ...state,
        type: {
          ...state.type,
          items: filterTypeItems,
          isDisabled: allItemsSelected,
        },
        items: newItems,
      };
      if (deletedFilterType === THING_FILTER_TYPE) {
        return {
          ...nextState,
          thingFilter: thingFilterInitialState,
        };
      } else if (deletedFilterType === DATE_FILTER_TYPE) {
        return {
          ...nextState,
          dateFilter: dateFilterInitialState,
        };
      }
      return nextState;
    }
    case RESET:
      return initialState;
    default:
      return {
        ...state,
        thingFilter: thingFilterReducer(state.thingFilter, action),
        dateFilter: dateFilterReducer(state.dateFilter, action),
      };
  }
};
