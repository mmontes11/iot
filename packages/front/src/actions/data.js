import {
  DATA_UPDATED,
  DATA_REQUEST,
  DATA_REQUEST_SUCCESS,
  DATA_REQUEST_ERROR,
  ADD_DATA_ITEM,
  RESET_DATA,
} from "constants/actionTypes/data";
import { EVENT_TYPE, MEASUREMENT_TYPE } from "constants/observationTypes";
import iotClient from "lib/iotClient";
import { SocketController } from "helpers/socketController";
import { THING_FILTER_TYPE, DATE_FILTER_TYPE } from "constants/filterTypes";
import { RESET } from "constants/actionTypes/common";
import * as fromState from "reducers";
import { TYPE, OBSERVATION, GROUPBY, THING } from "constants/params";

const getParams = (filterItems, observation, groupBy, thing, timePeriod, startDate, endDate) => {
  let params = {
    type: observation,
  };
  if (groupBy) {
    params = {
      ...params,
      groupBy,
    };
  }
  if (filterItems.includes(THING_FILTER_TYPE) && thing) {
    params = {
      ...params,
      thing,
    };
  }
  if (filterItems.includes(DATE_FILTER_TYPE)) {
    if (timePeriod) {
      params = {
        ...params,
        timePeriod,
      };
    } else {
      if (startDate) {
        params = {
          ...params,
          startDate: startDate.toISOString(),
        };
      }
      if (endDate) {
        params = {
          ...params,
          endDate: endDate.toISOString(),
        };
      }
    }
  }
  return params;
};

const requestStats = (
  filterItems,
  type,
  observation,
  thing,
  timePeriod,
  startDate,
  endDate,
  onStart,
  onSuccess,
  onError,
) => {
  const params = getParams(filterItems, observation, null, thing, timePeriod, startDate, endDate);
  if (type === EVENT_TYPE) {
    onStart();
    iotClient.eventService
      .getStats(params)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else if (type === MEASUREMENT_TYPE) {
    onStart();
    iotClient.measurementService
      .getStats(params)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else {
    onError(new Error("Invalid type"));
  }
};

export const getStats = () => (dispatch, getState) => {
  const state = getState();
  const filterItems = state.filters.items;
  const type = fromState.getParam(state, TYPE).selectedItem;
  const observation = fromState.getParam(state, OBSERVATION).selectedItem;
  const thing = state.filters.thingFilter.selectedItem;
  const {
    timePeriod: { selectedItem: selectedTimePeriod },
    custom: { startDate, endDate },
  } = state.filters.dateFilter;
  if (!type || !observation) {
    return;
  }
  requestStats(
    filterItems,
    type,
    observation,
    thing,
    selectedTimePeriod,
    startDate,
    endDate,
    () => dispatch({ type: DATA_REQUEST }),
    res => {
      dispatch({ type: DATA_REQUEST_SUCCESS });
      dispatch({ type: DATA_UPDATED, items: res.body.stats });
    },
    () => {
      dispatch({ type: DATA_REQUEST_ERROR });
      dispatch({ type: RESET, preserveError: true });
    },
  );
};

const requestData = (
  filterItems,
  type,
  observation,
  groupBy,
  thing,
  timePeriod,
  startDate,
  endDate,
  onStart,
  onSuccess,
  onError,
) => {
  const params = getParams(filterItems, observation, groupBy, thing, timePeriod, startDate, endDate);
  if (type === EVENT_TYPE) {
    onStart();
    iotClient.eventService
      .getData(params)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else if (type === MEASUREMENT_TYPE) {
    onStart();
    iotClient.measurementService
      .getData(params)
      .then(res => onSuccess(res))
      .catch(err => onError(err));
  } else {
    onError(new Error("Invalid type"));
  }
};

export const getData = () => (dispatch, getState) => {
  const state = getState();
  const filterItems = state.filters.items;
  const type = fromState.getParam(state, TYPE).selectedItem;
  const observation = fromState.getParam(state, OBSERVATION).selectedItem;
  const groupBy = fromState.getParam(state, GROUPBY).selectedItem;
  const thing = state.filters.thingFilter.selectedItem;
  const {
    timePeriod: { selectedItem: selectedTimePeriod },
    custom: { startDate, endDate },
  } = state.filters.dateFilter;
  if (!type || !observation) {
    return;
  }
  requestData(
    filterItems,
    type,
    observation,
    groupBy,
    thing,
    selectedTimePeriod,
    startDate,
    endDate,
    () => dispatch({ type: DATA_REQUEST }),
    ({ body: { data, things } }) => {
      dispatch({ type: DATA_REQUEST_SUCCESS });
      dispatch({ type: DATA_UPDATED, items: data, things });
    },
    () => {
      dispatch({ type: DATA_REQUEST_ERROR });
      dispatch({ type: RESET, preserveError: true });
    },
  );
};

let socketController = null;

export const finishRealTimeData = () => () => {
  if (socketController !== null) {
    socketController.close();
  }
};

export const startRealTimeData = () => (dispatch, getState) => {
  const state = getState();
  const thing = fromState.getParam(state, THING).selectedItem;
  const type = fromState.getParam(state, TYPE).selectedItem;
  if (!thing || !type) {
    return;
  }
  finishRealTimeData()();
  const onData = data => {
    const {
      data: { items },
    } = getState();
    if (items.length === 0) {
      dispatch({ type: DATA_REQUEST_SUCCESS });
    }
    dispatch({ type: ADD_DATA_ITEM, item: data });
  };
  const onError = error => {
    dispatch({ type: DATA_REQUEST_ERROR, error });
    dispatch({ type: RESET, preserveError: true });
  };
  socketController = new SocketController(thing, type, onData, onError);
  socketController.listen();
  dispatch({ type: DATA_REQUEST });
};

export const resetData = () => dispatch => {
  dispatch({ type: RESET_DATA });
};
