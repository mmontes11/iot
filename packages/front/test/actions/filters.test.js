import { selectFilterType, addFilterType, deleteFilterType } from "actions/filters";

describe("actions/filters", () => {
  it("dispatches a selectFilterType action", () => {
    const thunk = selectFilterType();
    const dispatch = jest.fn();
    thunk(dispatch);
    expect(thunk).toBeAThunk();
    expect(thunk).toMatchSnapshot();
  });
  it("dispatches a addFilterType action", () => {
    const thunk = addFilterType();
    const dispatch = jest.fn();
    thunk(dispatch);
    expect(thunk).toBeAThunk();
    expect(thunk).toMatchSnapshot();
  });
  it("dispatches a deleteFilterType action", () => {
    const thunk = deleteFilterType();
    const dispatch = jest.fn();
    thunk(dispatch);
    expect(thunk).toBeAThunk();
    expect(thunk).toMatchSnapshot();
  });
});
