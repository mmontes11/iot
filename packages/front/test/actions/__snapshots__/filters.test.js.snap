// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions/filters dispatches a addFilterType action 1`] = `[Function]`;

exports[`actions/filters dispatches a deleteFilterType action 1`] = `[Function]`;

exports[`actions/filters dispatches a selectFilterType action 1`] = `[Function]`;
